### PR TITLE
Globally cache single TexManager instances.

### PR DESCRIPTION
**Pull request by anntzer**
This allows sharing its caches across renderer instances.

(If it was up to me this classes would be replaced by module-level
functions and a module-level cache, but heh.)

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by timhoffm** [comment]
`TexManager` carries some state, e.g. `font_family`. Do we get into inconsistencies when changing e.g. `rcParams['font.family']` and then reusing the existing instance?

**Comment by timhoffm** [comment]
Hm, losts of magic.

This would be ok then, but hard to understand. What's the gain of having just a single instance and is it worth the complexity?

**Comment by anntzer** [comment]
texmanager caches its results (see rgba_arrayd, grey_arrayd) but right now this caching is only used for a single renderer; the next renderer call (e.g. if one saves to pdf or svg) needs to redo all that computation.  Always returning the same instance allows reusing that cache.

I initially thought that textpath did the same, but apparently it doesn't, so I guess we don't need to do that; undoing that part.

File: lib/matplotlib/textpath.py [hunk 1]
```diff
@@ -25,7 +25,9 @@ def _get_adobe_standard_encoding():
 
 class TextToPath(object):
     """
-    A class that convert a given text to a path using ttf fonts.
+    A class that converts strings to paths.
+
+    Repeated calls to this constructor always return the same instance.
```
Comment (QuLogic) [inline lib/matplotlib/textpath.py]: This is not true now?

**Comment by tacaswell** [comment]
> Hm, losts of magic.

There are two hard problems in programming: naming things, cache invalidation, and off-by-one bugs.

File: lib/matplotlib/textpath.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/textpath.py]: indeed, fixed

**Comment by anntzer** [comment]
Previously each canvas would create its own texmanager instance when it tried to invoke usetex, but now they all use the same instance (well, for that part you can check that the `lru_cache` pattern is used elsewhere in the codebase for the same purpose).
The fact that the tests still pass (and the tests test usetex more than once...) shows that there are no adverse effects.

**Comment by timhoffm** [comment]
So, the only advantage is saving computation time, when two renderes render the same figure (e.g. first to screen and then to PNG)?

If that‘s the case, I‘m -0.5 on this. The gain seems limited and the code logic gets more complex (also I‘m still not 100% sure that reusing the cache cannot lead to troubles. OTOH the code of this change is simple enough, so If others feel it‘s worth doing, Iöm fine with that.

**Comment by anntzer** [comment]
This would also allow not trying to cache TexManager instances at other places; e.g. right now contour handling code caches TexManager instances to figure out the size of tex strings (https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/contour.py#L269); with this PR we could just recreate TexManager on the fly when needed (as we'd always get the same instance).
In other words, the (minor, IMO) additional complexity here is paid back by simplifications elsewhere.

**Comment by efiring** [comment]
Logically, this should be a singleton, as far as I can see.  I see only 2 caches.  One is the on-disk long-term-persistent cache of pngs etc., which is already global, not instance-specific.  The other is _rc_cache, which is just a tiny dictionary.  Therefore I don't see how making this a singleton is actually saving much more than the time it takes to make the instance.  Am I missing something?  I'm inclined to merge it solely on the logical grounds that it *should* be a singleton; having more than one instance looks pointless.  (And I agree that having it as a class is probably also unnecessary, but it looks harmless.)

**Comment by anntzer** [comment]
Oh, actually I missed the fact that rgba_arrayd and grey_arrayd are class-level caches so are shared across instances, so my argument is not really strong :p
I guess the main point is that we should not be bothering with e.g. the get_texmanager() methods on the backends and TextPath then...

**Comment by jklymak** [comment]
I'll hold off on merging pending further comment from @efiring

**Comment by efiring** [comment]
Maybe we can discuss this briefly on Monday. I'm wondering whether it would make more sense to just make a single instance in the texmanager module, and use that directly elsewhere.  Perhaps the argument against that is the instantiation cost it incurs at startup, regardless of whether the TexManager instance is ever actually needed.  In addition, it looks to me like it would fail with a RuntimeError on Google App Engine.  It looks like someone started to put in a fix for that, but it wasn't completed.  So most likely I am just misunderstanding. 

Another question: in this PR, when the cached instance is returned by __new__, __init__ is still called, isn't it?  If so, we lose a little performance if we try to eliminate the get_texmanager methods.  This would not be the case if just made a single instance in the texmanager module, and always used that directly.

To avoid the __init__ overhead, wouldn't one need to move all of the __init__ code into __new__, and make __init__ a no-op?

**Comment by anntzer** [comment]
> Maybe we can discuss this briefly on Monday. I'm wondering whether it would make more sense to just make a single instance in the texmanager module, and use that directly elsewhere. Perhaps the argument against that is the instantiation cost it incurs at startup, regardless of whether the TexManager instance is ever actually needed. In addition, it looks to me like it would fail with a RuntimeError on Google App Engine. It looks like someone started to put in a fix for that, but it wasn't completed. So most likely I am just misunderstanding.

GAE is basically not supported, see e.g. references in https://github.com/matplotlib/matplotlib/pull/8939.  Will remove "pretense" of supporting GAE in texmanager in a separate PR.

> Another question: in this PR, when the cached instance is returned by new, init is still called, isn't it? If so, we lose a little performance if we try to eliminate the get_texmanager methods. This would not be the case if just made a single instance in the texmanager module, and always used that directly.
> To avoid the init overhead, wouldn't one need to move all of the init code into new, and make init a no-op?

Yes, my mistake, good catch. Fixed it.

File: lib/matplotlib/texmanager.py [hunk 2]
```diff
@@ -146,6 +154,8 @@ def __init__(self):
         self._font_preamble = '\n'.join(
             [r'\usepackage{type1cm}'] + cmd + [r'\usepackage{textcomp}'])
 
+        return self
```
Comment (efiring) [inline lib/matplotlib/texmanager.py]: Why is `_reinit` returning `self`?  I don't see the return value being used anywhere.

Comment (anntzer) [reply lib/matplotlib/texmanager.py]: Leftover of when I tried to put everything in `__new__`... sorry for the mess.

**Pull request merged**